{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b53501c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl\n",
    "from needletail import parse_fastx_file, NeedletailError, reverse_complement, normalize_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13bbf995",
   "metadata": {},
   "outputs": [],
   "source": [
    "ASESMBLY_FASTA_FILE = \"../a9_genome_masked.fa\"\n",
    "MIN_LENGTH = 10_000_000\n",
    "OUTPUT_FILENAME = \"hoiho.karyotype.txt\"\n",
    "TOTAL_COLORS = 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9026339d",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_lengths = {}\n",
    "\n",
    "try:\n",
    "    for record in parse_fastx_file(ASESMBLY_FASTA_FILE):\n",
    "        chr_lengths[record.id] = len(record.seq)\n",
    "except NeedletailError:\n",
    "    print(\"Invalid Fastq file\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6db67cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# How many are >= MIN_LENGTH\n",
    "long_chrs = {k: v for k, v in chr_lengths.items() if v >= MIN_LENGTH}\n",
    "print(f\"Number of chromosomes >= {MIN_LENGTH:,} bp: {len(long_chrs)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64ddece3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get max, mean of long_chrs\n",
    "max_length = max(long_chrs.values())\n",
    "mean_length = sum(long_chrs.values()) / len(long_chrs) if long_chrs else 0\n",
    "total_length = sum(long_chrs.values())\n",
    "print(f\"Max length: {max_length:,} bp\")\n",
    "print(f\"Mean length: {mean_length:,.2f} bp\")\n",
    "print(f\"Total length: {total_length:,} bp\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "953e4e4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Order by length\n",
    "ordered_chrs = dict(sorted(long_chrs.items(), key=lambda item: item[1], reverse=True))\n",
    "# Output in the following format\n",
    "#\n",
    "# chr - S1 1 0 155644563 c1\n",
    "# chr - S2 2 0 123023803 c2\n",
    "# chr - S3 3 0 101869369 c3\n",
    "# chr - S4 4 0 89082643 c4\n",
    "# chr - S5 5 0 86460390 c5\n",
    "# etc..\n",
    "\n",
    "with open(OUTPUT_FILENAME, \"w\") as f:\n",
    "    for i, (chr_name, length) in enumerate(ordered_chrs.items(), start=1):\n",
    "        f.write(f\"chr - {chr_name} {i} 0 {length} c{i % TOTAL_COLORS}\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "default",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
