{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import polars as pl\n",
    "import plotly.express as px\n",
    "\n",
    "import plotly.io as pio\n",
    "\n",
    "pio.renderers.default = \"plotly_mimetype+notebook_connected\"\n",
    "\n",
    "# Load up metadata from \"DNA from hoiho genomesv2.csv\"\n",
    "metadata = pl.read_csv(\n",
    "    \"Hoiho_Genomes_Cleaned_Nov4_24_24.csv\", separator=\"\\t\"\n",
    ")\n",
    "\n",
    "metadata = metadata.with_columns(\n",
    "    pl.col(\"ID\").replace(\"P29 \", \"P29\").alias(\"ID\")\n",
    ")\n",
    "\n",
    "# Next is C101/CE9\n",
    "metadata = metadata.with_columns(\n",
    "    pl.col(\"ID\").replace(\"C101/CE9\", \"CE9\").alias(\"ID\")\n",
    ")\n",
    "\n",
    "# bcftools view '-eF_MISSING>0.2 | QUAL < 99' -s^PP6,PP19 -v snps --threads 32 --min-af 0.05:minor -m2 -M2  -Ob -o merged.a9.filtered.qual99_fmissing0.2.maf0.05.biallelic.bcf merged.a9.unfiltered.bcf\n",
    "# pixi run plink2 --allow-extra-chr --bcf merged.a9.filtered.qual99_fmissing0.2.maf0.05.biallelic.bcf --pca --out merged.a9.filtered.qual99_fmissing0.2.maf0.05.biallelic --vcf-half-call missing\n",
    "\n",
    "# bcftools view '-eQUAL < 99' -s^PP6,PP19 -v snps --threads 32 --min-af 0.05:minor -m2 -M2  -Ob -o merged.a9.filtered.qual99.maf0.05.biallelic.bcf merged.a9.unfiltered.bcf\n",
    "# pixi run plink2 --allow-extra-chr --bcf merged.a9.filtered.qual99.maf0.05.biallelic.bcf --pca --out merged.a9.filtered.qual99.maf0.05.biallelic --vcf-half-call missing\n",
    "\n",
    "eigenvec = pl.read_csv(\"merged.a9.filtered.qual99_fmissing0.2.maf0.05.biallelic.eigenvec\",\n",
    "  separator=\"\\t\")\n",
    "# eigenvec = pl.read_csv(\"merged.a9.filtered.qual99.maf0.05.biallelic\", separator=\"\\t\")\n",
    "eigenvec = pl.read_csv(\"removed_qual99_maf0.05.eigenvec\", separator=\"\\t\")\n",
    "eigenvec = eigenvec.rename(mapping={\"#IID\": \"ID\"})\n",
    "joined = eigenvec.join(metadata, on=\"ID\", how=\"inner\")\n",
    "\n",
    "joined = joined.to_pandas()\n",
    "\n",
    "eigenvec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PC2 is split by sex, males are positive, females are negative \n",
    "# Create a field called \"Sex_Genome\"\n",
    "\n",
    "# Create a field\n",
    "joined['Sex_Genome'] = joined['PC2'] > 0\n",
    "\n",
    "# Convert to Male/Female\n",
    "joined['Sex_Genome'] = joined['Sex_Genome'].map({True: 'Male', False: 'Female'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save dataframe with ID and Sex_Genome\n",
    "joined[['ID', 'Sex_Genome', 'Sex']].to_csv(\"GeneticSex.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "joined.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "default",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
